### ISSUE TITLE
Keycloak quarkus: NullPointerException when using EmailTemplateProvider to send email in custom SPI

### ISSUE THREAD

**Issue by letronghieu4897** [keycloak/keycloak#23162]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

dist/quarkus

### Short information
- Keycloak running on container.
- Custom SPI.
- Brute-force protection.
- Sending email.

### Description
We custom Bruce-force protector SPI to add an action whenever account was disabled by brute-force attack we should send email to notify for the user of that account.

### My Code
```
public class CustomBruteForceProtector extends DefaultBruteForceProtector
{
	private static final Logger logger = Logger.getLogger(CustomBruteForceProtector.class);

	public CustomBruteForceProtector(KeycloakSessionFactory factory)
	{
		super(factory);
	}

	public static class Factory extends DefaultBruteForceProtectorFactory
	{

		private CustomBruteForceProtector protector;

		@Override
		public BruteForceProtector create(KeycloakSession session)
		{
			return protector;
		}

		@Override
		public void postInit(KeycloakSessionFactory factory)
		{
			protector = new CustomBruteForceProtector(factory);
			protector.start();
		}

		@Override
		public void close()
		{
			protector.shutdown();
		}
	}

	@Override
	protected void failure(KeycloakSession session, LoginEvent event)
	{
		super.failure(session, event);

		final RealmModel realm = getRealmModel(session, event);
		final UserLoginFailureModel userLoginFailure = getUserModel(session, event);
		final UserModel user = session.users().getUserById(realm, userLoginFailure.getUserId());
		if (super.isTemporarilyDisabled(session, realm, user))
		{
			try
			{
				Map<String, Object> attributes = new HashMap<>();
				attributes.put("user", user);

				KeycloakSession newSession = factory.create();
				newSession.getContext().setRealm(realm);

				EmailTemplateProvider emailTemplateProvider = newSession.getProvider(EmailTemplateProvider.class);
				emailTemplateProvider.setRealm(realm);
				emailTemplateProvider.setUser(user);
				emailTemplateProvider.send("bruteForceSubject", new ArrayList<>(), "brute-force.ftl", attributes);
			}
			catch (Exception e)
			{
				logger.error("Failed to send email", e);
			}
		}
	}
}
```

### Version

20.0.2

### Expected behavior

Send email successfully.

### NullPointerException

org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.processTemplate(FreeMarkerEmailTemplateProvider.java:246)
org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.send(FreeMarkerEmailTemplateProvider.java:261)
org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.send(FreeMarkerEmailTemplateProvider.java:256)
com.fyndoo.direct.portal.CustomBruteForceProtector.failure(CustomBruteForceProtector.java:83)
org.keycloak.services.managers.DefaultBruteForceProtector.run(DefaultBruteForceProtector.java:221)
java.base/java.lang.Thread.run(Thread.java:829)
**Caused by: java.lang.NullPointerException**
org.keycloak.quarkus.runtime.hostname.**DefaultHostnameProvider.getRealmFrontEndUrl(DefaultHostnameProvider.java:176)**
org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider.fromFrontEndUrl(DefaultHostnameProvider.java:148)
org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider.getScheme(DefaultHostnameProvider.java:76)
org.keycloak.models.KeycloakUriInfo.<init>(KeycloakUriInfo.java:46)
org.keycloak.services.DefaultKeycloakContext.getUri(DefaultKeycloakContext.java:78)
org.keycloak.services.DefaultKeycloakContext.getUri(DefaultKeycloakContext.java:85)
org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.processTemplate(FreeMarkerEmailTemplateProvider.java:215)

### How to Reproduce?

Custom brute-force protector, as per the provided code, enables brute-force protection. Subsequently, login attempts will continue to fail until the allowed limit for the number of failed login attempts is reached.

**Comment by vmuzikar** [comment]
This sounds more like a Core issue, reassigning for now.

**Comment by rmartinc** [comment]
@letronghieu4897 I think the problem is that the email needs a session with the URI information (created by a request), and the default brute force protector is a singleton that creates internally the sessions.

Have you tried creating the provider every time in the factory instead of using a singleton?

Something like the following:

```
    // in the factory
    @Override
    public BruteForceProtector create(KeycloakSession session) {
        return new CustomBruteForceProtector(session);
    }

    // in the CustomBruteForceProtector
    public CustomBruteForceProtector(KeycloakSession session) {
        super(session.getKeycloakSessionFactory());
        this.session = session;
    }
```

And use the passed session to send the emails instead creating a new one.

Let me know if that works for you.

**Comment by rmartinc** [comment]
Yesterday I realized there is a similar try to use emails with brute force detection in this [PR](https://github.com/keycloak/keycloak/pull/8679). It seems that the brute force protector is using its own background task for performance reasons and adding things to it can be complicated. So take care, here there are some specifics you have to consider and it can be complicated. The PR mentioned can be helpful for understanding the code.

**Comment by letronghieu4897** [comment]
> @letronghieu4897 I think the problem is that the email needs a session with the URI information (created by a request), and the default brute force protector is a singleton that creates internally the sessions.
> 
> Have you tried creating the provider every time in the factory instead of using a singleton?
> 
> Something like the following:
> 
> ```
>     // in the factory
>     @Override
>     public BruteForceProtector create(KeycloakSession session) {
>         return new CustomBruteForceProtector(session);
>     }
> 
>     // in the CustomBruteForceProtector
>     public CustomBruteForceProtector(KeycloakSession session) {
>         super(session.getKeycloakSessionFactory());
>         this.session = session;
>     }
> ```
> 
> And use the passed session to send the emails instead creating a new one.
> 
> Let me know if that works for you.

Thank you for that information. I have tried but it didn't work for me.

**Comment by rmartinc** [comment]
@letronghieu4897 Yesterday I spent some time on this and the problem is that the Brute Force Protector is using another thread to manage the events (using a queue). In that thread all the resteasy information is missing (as I think resteasy uses a ThreadLocal to store the data), among them the `KeycloakSession` and the `HttpRequest` are null and that produces the NPE. I could make it work doing something like the following (not very pretty but working):

```java
    // prepare the session for the email provider
    session.getContext().setRealm(realm);
    HttpRequest httpRequest = new HttpRequestImpl(MockHttpRequest.create("GET", frontendUri, frontendUri));
    ResteasyContext.getContextDataMap().put(HttpRequest.class, httpRequest);
    ResteasyContext.getContextDataMap().put(KeycloakSession.class, session);
    // send the email
    // ...
    Resteasy.clearContextData(); // better in finally
```

I see the email provider gets the URI from the request to attach the keycloak base URI for links. But what surprises me is that any new session created in a separated thread will fail on Resteasy data in general.

@pedroigor Am I right? I know that brute force is an internal provider, not thought to be extended, but maybe the issue can happen in another parts. Some providers will fail as soon as they are executed off-thread.

**Comment by pedroigor** [comment]
Yeah, whenever you are running code in a separate thread you need to share contextual data with the new thread. Some providers might rely on contextual data (realm, client, etc) but those are only available from the original thread. It is not an issue but how they were designed/implemented.

I'm not sure why but you are creating another session and not closing it at the end.

I would avoid extending this provider and leverage the event introduced here to solve the same use case https://github.com/keycloak/keycloak/issues/8769.

**Comment by rmartinc** [comment]
@pedroigor But my point is that if someone wants to send an email from the new USER_DISABLED_BY_PERMANENT_LOCKOUT event, the same issue is going to happen, right? The session is created by the brute-force class off-thread and then sending email is going to fail. Or am I missing something? Maybe we should fix that part at brute-force level.

Besides we also need an event to  USER_TEMPORAILY_DISABLED_BY_LOCKOUT (or similar) in order to perform what this extension was trying. But that's another story.

**Comment by pedroigor** [comment]
You can still invoke the `EmailSenderProvider` directly to avoid going through our templates. The key problem here is that processing templates requires access to providers that rely on contextual data. It is hard to know in advance which ones can run detached from the request thread. It is also hard to know which data can work and be shared across threads.

Sorry for not being clear, my point is that we should use a similar approach to that other issue to enable the `USER_TEMPORAILY_DISABLED_BY_LOCKOUT` event.
